### ISSUE TITLE
how to finetune in my own dataset?

### ISSUE THREAD

**Issue by beizhengren** [jfzhang95/pytorch-deeplab-xception#114]
Hi,
I want to fine tune in my own dataset, which including 3 classes. 
And, now I have the dataset with VOC style, but I don't know how to modify the structure of the net.
Can anyone help me?
Thank you

**Comment by HongChow** [comment]
> Hi,
> I want to fine tune in my own dataset, which including 3 classes.
> And, now I have the dataset with VOC style, but I don't know how to modify the structure of the net.
> Can anyone help me?
> Thank you

Have you done ?

**Comment by jaemin93** [comment]
Hi, @HongChow @beizhengren .
I can make to work my own dataset. I can help you.

1. dataloader/dataset/pascal.py : line 14
NUM_CLASSES = ?  

2. dataloader/__init__.py: line 9~11
maybe you don't use sbd, so just delete.

3. dataloader/utils.py: line25, line 91~
line 25 is NUM_CLASSES.
you modify return array.
ex) i have 3 grayscale label: 0, 1, 2
0: background, 1: dog, 2: cat
and you express class to RGB. black is background, red is dog, blue is cat.
background: [0, 0, 0] dog: [255, 0, 0], cat: [0, 0, 255]
so index of array is grayscale label. and RGB of array element is you want to express class.

return np.asarray[[0,0,0],[255,0,0],[0,0,255]]

4. you can run this command style.
if you have 2-gpu
CUDA_VISIBLE_DEVICES=0,1 python train.py --backbone xception \
--lr 0.007 \
--ft \
--workers 1 \
--epochs 50 \
--start_epoch 25 \
--batch-size 6 \
--gpu-ids 0,1 \
--checkname deeplab-xception \
--eval-interval 1 \
--dataset pascal

i hope you work it.

**Comment by beizhengren** [comment]
Hi,@jaemin93 
I'll try it.
Thank you for helping me!

**Comment by XUYUNYUN666** [comment]
> Hi, @HongChow @beizhengren .
> I can make to work my own dataset. I can help you.
> 
> 1. dataloader/dataset/pascal.py : line 14
>    NUM_CLASSES = ?
> 2. dataloader/**init**.py: line 9~11
>    maybe you don't use sbd, so just delete.
> 3. dataloader/utils.py: line25, line 91~
>    line 25 is NUM_CLASSES.
>    you modify return array.
>    ex) i have 3 grayscale label: 0, 1, 2
>    0: background, 1: dog, 2: cat
>    and you express class to RGB. black is background, red is dog, blue is cat.
>    background: [0, 0, 0] dog: [255, 0, 0], cat: [0, 0, 255]
>    so index of array is grayscale label. and RGB of array element is you want to express class.
> 
> return np.asarray[[0,0,0],[255,0,0],[0,0,255]]
> 
> 1. you can run this command style.
>    if you have 2-gpu
>    CUDA_VISIBLE_DEVICES=0,1 python train.py --backbone xception 
>    --lr 0.007 
>    --ft 
>    --workers 1 
>    --epochs 50 
>    --start_epoch 25 
>    --batch-size 6 
>    --gpu-ids 0,1 
>    --checkname deeplab-xception 
>    --eval-interval 1 
>    --dataset pascal
> 
> i hope you work it.

I want to know the performance trained on your dataset? I trained on my dataset, but it is poor, I am looking forward to your reply.

**Comment by jaemin93** [comment]
@XUYUNYUN666 
Have you tried the use-balanced option? If you have imbalance data, try this option. just add '--use-balanced-weights' cmd line.

**Comment by XUYUNYUN666** [comment]
> @XUYUNYUN666
> Have you tried the use-balanced option? If you have imbalance data, try this option. just add '--use-balanced-weights' cmd line.

No, I have not tried the use-balanced option. But now the performance is poor, I do not the reason, the miou is only around 0.5, can you help me , thank you very much.

**Comment by jaemin93** [comment]
> > @XUYUNYUN666
> > Have you tried the use-balanced option? If you have imbalance data, try this option. just add '--use-balanced-weights' cmd line.
> 
> No, I have not tried the use-balanced option. But now the performance is poor, I do not the reason, the miou is only around 0.5, can you help me , thank you very much.

I didn't get good performance, but there was a data imbalance issue, so I used that option and saw a big improvement in miou.
For example, suppose you have your own animal data set: 20 pictures of cats in your pictures, and 100 pictures of dogs.

**Comment by XUYUNYUN666** [comment]
> > > @XUYUNYUN666
> > > Have you tried the use-balanced option? If you have imbalance data, try this option. just add '--use-balanced-weights' cmd line.
> > 
> > 
> > No, I have not tried the use-balanced option. But now the performance is poor, I do not the reason, the miou is only around 0.5, can you help me , thank you very much.
> 
> I didn't get good performance, but there was a data imbalance issue, so I used that option and saw a big improvement in miou.
> For example, suppose you have your own animal data set: 20 pictures of cats in your pictures, and 100 pictures of dogs.

Thank you very much, but it seem that there is no the imbalance issue in my dataset.  I will try another backbone, maybe the so deep resnet is not suitable for my simple dataset. Except the deeplabv3+, have you tried other networks? Can you recommend some one to me?

**Comment by jaemin93** [comment]
> > > > @XUYUNYUN666
> > > > Have you tried the use-balanced option? If you have imbalance data, try this option. just add '--use-balanced-weights' cmd line.
> > > 
> > > 
> > > No, I have not tried the use-balanced option. But now the performance is poor, I do not the reason, the miou is only around 0.5, can you help me , thank you very much.
> > 
> > 
> > I didn't get good performance, but there was a data imbalance issue, so I used that option and saw a big improvement in miou.
> > For example, suppose you have your own animal data set: 20 pictures of cats in your pictures, and 100 pictures of dogs.
> 
> Thank you very much, but it seem that there is no the imbalance issue in my dataset. I will try another backbone, maybe the so deep resnet is not suitable for my simple dataset. Except the deeplabv3+, have you tried other networks? Can you recommend some one to me?

I have applied several papers to the encoder and decoder sections, but did not get better performance. I agree to change the backbone. If you ever find a better backbone, please share it. I think I'll apply Inception-Resnet V2.
